### PR TITLE
Improvements for UNIX A.out Loader

### PR DESCRIPTION
**Pull request by boricj**
As mentioned in #1, another set of improvements for this loader. I've cleaned up, improved and modeled the loader against the ELF loader as a reference.

Among other things:

- The a.out file structures are loaded into memory (as overlays if needed) ;
- The a.out file structures are marked up in the listing view (header, relocation tables, symbol table, string table) ;
- The relocation table is filled out ;
- Undefined symbols are allocated from an `EXTERNAL` memory block (just like for ELF).

It's not quite there yet, I have a couple of relocations/symbols bugs with the Slackware 2.3 C standard library I haven't nailed down and there's some cleaning up left to do. Therefore, I'm marking this as a draft for the time being.

One additional area of improvement would be adding stabs support, but that's probably best done inside a dedicated analyzer (like how DWARF support is implemented inside Ghidra) and out of scope of this loader. I haven't decided yet if I'll write one, that'll depend on where my own reverse-engineering/investigation efforts will lead me.

Like in my first PR, I'm fine with squashing.

### CODE REVIEW

**Comment by boricj** [comment]
I think it's in a good enough state now. There's still some weird stuff going on with some symbols whose addresses are beyond the end of their section, but I ran out of patience wrestling with that ancient junk, so meh.

I've deactivated the ability to load a.out binaries inside a program (`supportsLoadIntoProgram()`) because the refactoring broke that. I only care about static executables and relocatable objects (as a source for the version tracking tool) for now, so that's good enough for me. I don't know much about a.out shared libraries ; I can probably prepend section names with their file names when using `Add To Program...` in a separate PR if you need that (that might be enough for Linux a.out shared libraries when overriding the base load address), but I won't implement proper a.out shared library support in this loader myself anytime soon.

**Comment by colinbourassa** [comment]
Thanks! Since this is my first significant Ghidra contribution, I'm happy to defer to anyone with more knowledge about Ghidra coding style and architecture. It's nice to have it align better with the organization of the ELF loader.

If you could look into re-enabling `supportsLoadIntoProgram()`, that would be extremely helpful. My primary use case for A.out loading actually requires this. I'll merge this PR in the meantime.

**Comment by boricj** [comment]
> Thanks! Since this is my first significant Ghidra contribution, I'm happy to defer to anyone with more knowledge about Ghidra coding style and architecture. It's nice to have it align better with the organization of the ELF loader.

I'm not _that_ knowledgeable about Ghidra's internals, I know just enough to get into plenty of trouble (https://github.com/boricj/ghidra-delinker-extension). Hopefully these were steps in the right direction for upstreaming your a.out loader.

> If you could look into re-enabling `supportsLoadIntoProgram()`, that would be extremely helpful. My primary use case for A.out loading actually requires this. I'll merge this PR in the meantime.

Do you have documentation on a.out shared libraries somewhere? I can't seem to find anything useful online besides "it sucked hard in ye olden days so we switched to ELF decades ago". I've looked at Slackware 2.3's source code archives of glibc, ldso and Linux kernel and found some answers, but it's hard to piece together a coherent picture from that. What about other operating systems? I've read that the BSDs did it differently, but that's about it...

Also, what level of functionality are you looking for? If it's just loading Linux-style shared libraries by hand I can probably make that work, anything beyond that I probably can't help with.

**Comment by colinbourassa** [comment]
> Also, what level of functionality are you looking for? If it's just loading Linux-style shared libraries by hand I can probably make that work, anything beyond that I probably can't help with.

I don't exactly need support for a.out shared libraries, specifically -- just adding object files to the program being analyzed. My workflow looks like this:

* Import a VxWorks (monolith?) image using the generic binary file importer. Manually partition it into `.text` and `.data` with the VxWorks fixed/global address space.
* Run the `VxWorksSymTab_Finder` script to locate the named symbol table and use its contents to identify symbols throughout the image.
* Import a separate a.out object code file using `Add To Program...`. In my case, these files are identified in their headers as OMAGIC (0x107). During this step, the code in my original version of the a.out loader would find the symbols previously identified by `VxWorksSymTab_Finder`, and fix their addresses to point to the true location of each symbol. (The actual fixing was being done by `UnixAoutLoader::fixAddress()`.)

I found that fixing each address to point to the original symbol was very helpful, because then any change to a function signature (for example) would be automatically reflected in calls to that function from the code in the object file.

Would it be possible to restore this functionality? Please let me know if I'm thinking about this the wrong way.

**Comment by boricj** [comment]
To be honest, after giving it some thoughts I'm not even sure there is a right way to think about this.

Usually, Ghidra handles undefined symbols by defining an `EXTERNAL` block that holds placeholders and applied relocations against these, so that analysis can proceed as if everything's OK. Each dynamic library, if any, is loaded as a separate program so that they aren't duplicated into multiple programs that uses them.

As far as I can tell, loading libraries _into_ a program isn't the usual workflow. In-tree loaders that do support loading into programs mostly either add raw bits or add debugging symbols ; none of them appear to support loading libraries. However, some recent changes in the relocation table (the `Status` enum inside `Relocation.java` in particular) suggest that might change in the future, but I don't know to what extent.

Now, Linux-style a.out shared libraries already break that model because external symbols aren't undefined, every shared object was globally pre-allocated a virtual memory range and the relocations were pre-applied in advance. That means there are no symbols to put inside an `EXTERNAL` block or even symbols to mark as external ; these references will point to unallocated memory from Ghidra's point of view _unless_ that shared library is loaded, but that would duplicate these libraries into each program that uses them...

But there's also the problem that Ghidra really doesn't expect the relocation table to be touched after the initial load (my attempts to change that for unrelated reasons didn't reach consensus: https://github.com/NationalSecurityAgency/ghidra/pull/5010). That works out for Linux-style a.out dynamic linking because there are no relocations to process, but that's not the case for other operating systems. In theory, these shared libraries should be loaded alongside the initial program load so as not to break Ghidra's assumptions about its relocation table ; in practice that would be a mess with a.out because there are no standards for dynamic linking. Even worse, unlike ELF there's no metadata to tell what dynamic structures are where (the list of dynamic libraries to load on Linux being one example that I'm aware of) ; which leads to hacky heuristics like `VxWorksSymTab_Finder.java` to sort this mess out.

TL;DR I don't know. I'm not knowledgeable enough about Ghidra or a.out to figure it out on my own. I'll look into making loading a.out Linux-style shared libraries work since that at least doesn't break assumptions that Ghidra currently makes and it should be fairly straightforward to do. For the rest, it might be a good idea to open [a discussion on Ghidra's side](https://github.com/NationalSecurityAgency/ghidra/discussions) about this. Maybe I've missed something obvious, but I'm too confused right now to make sense of this.

**Comment by colinbourassa** [comment]
Thank you for your thoughts on this. My workflow is probably a little unusual in that the VxWorks image I'm loading contains code that would normally be shared library code (e.g. functions from `libc` or its equivalent). The main executive loop then dynamically loads a separate a.out OMAGIC file and passes control to a known function contained therein. The code in the OMAGIC file then references library symbols back in the original image.

As you suggest, it's probably worth a discussion thread back in the Ghidra project.
